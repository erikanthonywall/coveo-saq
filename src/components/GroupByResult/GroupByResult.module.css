.gbrContainer {
	margin: 16px;
}

.gbrHeader {
	display: flex;
	align-items: center;
	justify-content: space-between;
	border-bottom: 1px solid #fff;
    margin-bottom: 8px;
}

.collapseButton {
	cursor: pointer;
}

.collapseIcon {
	color: #fff;
	font-size: 24px;
	transition: transform 0.2s;
}

.collapseContainer {
	overflow: hidden;
    margin: 0 -8px;
    padding: 0 8px;
}

.gbrContainer.collapsed .collapseContainer {
	height: 0;
}

.gbrContainer.collapsed .collapseIcon {
	transform: rotate(180deg);
}

.gbrFieldName {
	color: #fff;
    font-size: 18px;
}

.gbrCheckLabel {
	display: flex;
	flex: 1;
	align-items: center;
	margin-bottom: 4px;
}

.gbrCheckLabelName {
	flex: 1;
	color: #fff;
	margin-left: 8px;
}

.filterRow {
	display: flex;
    align-items: center;
    cursor: pointer;
    border-radius: 16px;
    padding: 0px 8px;
    margin-left: -8px;
    margin-right: -8px;
}

.filterRow:hover {
	background-color: #848383;
}