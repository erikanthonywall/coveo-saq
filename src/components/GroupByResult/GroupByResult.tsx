import React, { useState } from 'react';
import styles from './GroupByResult.module.css';
import _indexOf from 'lodash/indexOf';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronUp } from '@fortawesome/free-solid-svg-icons';
import { IGroupByResult, IFilters } from '../../types';

interface IProps {
	/** Whether this Group By Result group is initially collapsed */
	isCollapsedInitial: boolean,

	/** The category of Group By Results to render */
	groupByResult: IGroupByResult,

	/** The set of currently selected Group By Result filters */
	filters: IFilters,

	/** The function to call to change the filter options */
	onChange(field: string, values: Array<string>): void
}

const GroupByResult: React.FC<IProps> = ({ isCollapsedInitial, groupByResult, filters, onChange }) => {
	const [isCollapsed, setIsCollapsed] = useState<boolean>(isCollapsedInitial);

	const onCheckFilter = (value: string, checked: boolean) => {
		if (checked) {
			onChange(groupByResult.field, [
				...filters[groupByResult.field],
				value
			]);
		} else {
			const i = _indexOf(filters[groupByResult.field], value);

			if (i >= 0) {
				onChange(groupByResult.field, [
					...filters[groupByResult.field].slice(0, i),
					...filters[groupByResult.field].slice(i + 1)
				]);
			}
		}
	}

	const isChecked = (value: string) => {
		return _indexOf(filters[groupByResult.field], value) >= 0;
	}

	if (groupByResult.values.length === 0) {
		return null;
	}

	return (
		<div className={`noselect ${styles.gbrContainer} ${isCollapsed ? styles.collapsed : ''}`}>
			<div className={styles.gbrHeader}>
				<h3 data-test-id="gbrFieldName" className={styles.gbrFieldName}>
					{groupByResult.label}
				</h3>

				<div
					className={styles.collapseButton}
					onClick={() => setIsCollapsed(!isCollapsed)}>
					
					<FontAwesomeIcon
						icon={faChevronUp}
						className={styles.collapseIcon} />
				</div>
			</div>

			<div className={styles.collapseContainer}>
				{
					groupByResult.values.map((v, i) =>
						<div key={i} className={styles.filterRow}>
							<label className={`checkbox ${styles.gbrCheckLabel}`}>
								<input
									type="checkbox"
									checked={isChecked(v.name)}
									onChange={(e) => onCheckFilter(v.name, e.target.checked)} />
								
								<span data-test-id="gbrCheckLabelName" className={styles.gbrCheckLabelName}>
									{v.name === 'true' ? 'On Sale' : v.name}
								</span>
							</label>
						</div>
					)
				}
			</div>
		</div>
	);
}

export default GroupByResult;