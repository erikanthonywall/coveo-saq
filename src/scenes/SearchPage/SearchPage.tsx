import React, { useState, useEffect } from 'react';
import _debounce from 'lodash/debounce';

import Header from '../../components/Header/Header';
import SideBar from '../../components/SideBar/SideBar';
import Search from '../../components/Search/Search';
import ProductList from '../../components/ProductList/ProductList';

import { searchProducts } from '../../api/products';
import { IProduct, IGroupByResult, IQueryResult, IFilters, ESortField, ESortDirection } from '../../types';

let firstResult: number = 0;
let appendResults = true;
let overwriteGroupByResults = true;

const executeSearch = (
	searchText: string,
	filters: IFilters,
	sortField: ESortField,
	sortDir: ESortDirection,
	setIsLoading: Function,
	parseResult: Function
) => {
	setIsLoading(true);

	searchProducts(searchText, filters, firstResult, sortField, sortDir).then((res) => {
		setIsLoading(false);
		parseResult(res);
	});
};

const debouncedSearch = _debounce(executeSearch, 1000);

const SearchPage: React.FC = () => {
	const [isSideBarVisible, setIsSideBarVisible] = useState<boolean>(false);
	const [searchText, setSearchText] = useState<string>('');
	const [isLoading, setIsLoading] = useState<boolean>(true);
	const [products, setProducts] = useState<Array<IProduct>>([]);
	const [groupByResults, setGroupByResults] = useState<Array<IGroupByResult>>([]);
	const [filters, setFilters] = useState<IFilters>(EMPTY_FILTERS);
	const [sortField, setSortField] = useState<ESortField>(ESortField.relevance);
	const [sortDirection, setSortDirection] = useState<ESortDirection>(ESortDirection.descending);

	// Handle searching when the search input is changed
	useEffect(() => {
		appendResults = false;
		firstResult = 0;
		overwriteGroupByResults = true;
		setFilters(EMPTY_FILTERS);
		debouncedSearch(searchText, filters, sortField, sortDirection, setIsLoading, parseResult);
		// eslint-disable-next-line
	}, [searchText]);

	// Handle searching when filter options or sorting are changed
	useEffect(() => {
		appendResults = false;
		firstResult = 0;
		debouncedSearch(searchText, filters, sortField, sortDirection, setIsLoading, parseResult);
		// eslint-disable-next-line
	}, [filters, sortField, sortDirection]);

	// Parse the results of the search and set them in state
	const parseResult = (res: IQueryResult) => {
		if (appendResults) {
			setProducts([
				...products,
				...res.results
			]);
		} else {
			setProducts(res.results);
		}

		appendResults = true;

		if (overwriteGroupByResults) {
			console.log(res.groupByResults);
			setGroupByResults(res.groupByResults);
			overwriteGroupByResults = false;
		}
	};

	// Load more results from the autoscroller
	const loadMoreResults = () => {
		firstResult += 12;
		executeSearch(searchText, filters, sortField, sortDirection, setIsLoading, parseResult);
	};

	const setFilterOptions = (field: string, values: Array<string>) => {
		setFilters({
			...filters,
			[field]: values
		});
	};

	const setSortingOptions = (field: ESortField) => {
		if (field === sortField) {
			if (sortDirection === ESortDirection.ascending) {
				setSortDirection(ESortDirection.descending);
			} else {
				setSortDirection(ESortDirection.ascending);
			}
		} else {
			setSortField(field);
			setSortDirection(ESortDirection.descending);
		}
	};

	return (
		<>
			<Header onShowSideBar={() => setIsSideBarVisible(!isSideBarVisible)}>
				<Search searchText={searchText} onChange={setSearchText} />
			</Header>

			<main className="main">
				<SideBar
					isMobileVisible={isSideBarVisible}
					groupByResults={groupByResults}
					sortField={sortField}
					sortDirection={sortDirection}
					onChangeSort={setSortingOptions}
					filters={filters}
					onChangeFilters={setFilterOptions} />

				<ProductList
					isLoading={isLoading}
					endOfResults={products.length <= firstResult}
					products={products}
					onRequestNextPage={loadMoreResults} />
			</main>
		</>
	);
}

export const EMPTY_FILTERS = {
	tpenspecial: [],
	tpdisponibilite: [],
	tpcategorie: [],
	tppays: [],
	tpmillesime: [],
	tpcepagenomsplitgroup: [],
	tpinventairenomsuccursalesplitgroup: [],
	tppastilledegout: [],
	tpfamilledevinsplitgroup: [],
	tpaccordsnommenu: [],
	tpobservationsgustativesacidite: [],
	tpobservationsgustativescorps: [],
	tpobservationsgustativessucre: [],
	tpobservationsgustativestannins: [],
	tpobservationsgustativestexture: []
};

export default SearchPage;