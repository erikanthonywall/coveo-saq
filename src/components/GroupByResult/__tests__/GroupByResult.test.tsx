import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import GroupByResult from '../GroupByResult';
import { EMPTY_FILTERS } from '../../../scenes/SearchPage/SearchPage';

describe('GroupByResult', () => {
	test('snapshot', () => {
		const component = renderer.create(
			<GroupByResult 
				isCollapsedInitial={false}
				filters={EMPTY_FILTERS}
				groupByResult={MOCK_GBR[2]}
				onChange={() => {}} />
		);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders the title correctly', () => {
		const wrapper = mount(
			<GroupByResult 
				isCollapsedInitial={false}
				filters={EMPTY_FILTERS}
				groupByResult={MOCK_GBR[2]}
				onChange={() => {}} />
		);

		expect(wrapper.find(`h3[data-test-id="gbrFieldName"]`).text()).toEqual('Category');
	});

	it('renders the filter options', () => {
		const wrapper = mount(
			<GroupByResult 
				isCollapsedInitial={false}
				filters={EMPTY_FILTERS}
				groupByResult={MOCK_GBR[2]}
				onChange={() => {}} />
		);

		expect(wrapper.find(`span[data-test-id="gbrCheckLabelName"]`).length).toEqual(10);
		expect(wrapper.find(`span[data-test-id="gbrCheckLabelName"]`).at(4).text()).toEqual('Rhum brun');
	});
});

const MOCK_GBR = [{
	"field": "tpenspecial",
	"label": "On Sale",
	"values": [{
		"count": 84,
		"name": "true"
	}]
}, {
	"field": "tpdisponibilite",
	"label": "Availability",
	"values": [{
		"count": 4573,
		"name": "En Ligne"
	}, {
		"count": 213,
		"name": "Bientôt disponible"
	}, {
		"count": 134,
		"name": "Commande Spéciale"
	}, {
		"count": 13155,
		"name": "En Succursale"
	}]
}, {
	"field": "tpcategorie",
	"label": "Category",
	"values": [{
		"count": 6348,
		"name": "Vin rouge"
	}, {
		"count": 3156,
		"name": "Vin blanc"
	}, {
		"count": 289,
		"name": "Vin rosé"
	}, {
		"count": 317,
		"name": "Champagne"
	}, {
		"count": 35,
		"name": "Rhum brun"
	}, {
		"count": 85,
		"name": "Whiskey américain"
	}, {
		"count": 433,
		"name": "Whisky écossais"
	}, {
		"count": 279,
		"name": "Vin mousseux"
	}, {
		"count": 27,
		"name": "Xérès"
	}, {
		"count": 180,
		"name": "Dry gin"
	}]
}, {
	"field": "tppays",
	"label": "Country",
	"values": [{
		"count": 5535,
		"name": "France"
	}, {
		"count": 2237,
		"name": "Italie"
	}, {
		"count": 1147,
		"name": "États-Unis"
	}, {
		"count": 93,
		"name": "Autriche"
	}, {
		"count": 232,
		"name": "Australie"
	}, {
		"count": 858,
		"name": "Espagne"
	}, {
		"count": 123,
		"name": "Allemagne"
	}, {
		"count": 541,
		"name": "Royaume Uni"
	}, {
		"count": 1002,
		"name": "Canada"
	}, {
		"count": 406,
		"name": "Portugal"
	}]
}, {
	"field": "tpmillesime",
	"label": "Vintage",
	"values": [{
		"count": 1856,
		"name": "2017"
	}, {
		"count": 2127,
		"name": "2016"
	}, {
		"count": 887,
		"name": "2018"
	}, {
		"count": 1616,
		"name": "2015"
	}, {
		"count": 983,
		"name": "2014"
	}, {
		"count": 292,
		"name": "2011"
	}, {
		"count": 209,
		"name": "2010"
	}, {
		"count": 425,
		"name": "2012"
	}, {
		"count": 104,
		"name": "2009"
	}, {
		"count": 513,
		"name": "2013"
	}]
}, {
	"field": "tpcepagenomsplitgroup",
	"label": "Variety",
	"values": [{
		"count": 1198,
		"name": "Syrah"
	}, {
		"count": 1382,
		"name": "Cabernet-sauvignon"
	}, {
		"count": 1518,
		"name": "Pinot noir"
	}, {
		"count": 1213,
		"name": "Merlot"
	}, {
		"count": 1619,
		"name": "Chardonnay"
	}, {
		"count": 430,
		"name": "Sangiovese"
	}, {
		"count": 160,
		"name": "Corvina"
	}, {
		"count": 131,
		"name": "Rondinella"
	}, {
		"count": 39,
		"name": "Molinara"
	}, {
		"count": 126,
		"name": "Shiraz"
	}]
}, {
	"field": "tpinventairenomsuccursalesplitgroup",
	"label": "Branch",
	"values": [{
		"count": 3367,
		"name": "Jonquière 33512"
	}, {
		"count": 2120,
		"name": "Galeries Charlesbourg 33616"
	}, {
		"count": 1352,
		"name": "Estimauville 33545"
	}, {
		"count": 1504,
		"name": "Valleyfield - Ch. Larocque 23071"
	}, {
		"count": 3461,
		"name": "Exclusivement pour les restaurateurs 23390"
	}, {
		"count": 2878,
		"name": "Châteauguay - Boul. d'Anjou 23129"
	}, {
		"count": 3250,
		"name": "Carrefour de la Pointe 23106"
	}, {
		"count": 3860,
		"name": "Exclusivement pour les restaurateurs 23385"
	}, {
		"count": 2089,
		"name": "Exclusivement pour les restaurateurs 33691"
	}, {
		"count": 963,
		"name": "Saint-Tite 33599"
	}]
}, {
	"field": "tppastilledegout",
	"label": "Flavour Profile",
	"values": [{
		"count": 296,
		"name": "Aromatique et charnu."
	}, {
		"count": 49,
		"name": "Corsé et complexe."
	}, {
		"count": 25,
		"name": "Agrumes."
	}, {
		"count": 89,
		"name": "Fruité et extra-doux."
	}, {
		"count": 286,
		"name": "Fruité et généreux."
	}, {
		"count": 293,
		"name": "Aromatique et souple."
	}, {
		"count": 23,
		"name": "Épicé."
	}, {
		"count": 19,
		"name": "Fruité."
	}, {
		"count": 136,
		"name": "Aromatique et rond."
	}, {
		"count": 22,
		"name": "Corsé et fumé."
	}]
}, {
	"field": "tpfamilledevinsplitgroup",
	"label": "Family of Wine",
	"values": [{
		"count": 1981,
		"name": "Sec"
	}, {
		"count": 898,
		"name": "Moyennement corsé"
	}, {
		"count": 289,
		"name": "Épicé"
	}, {
		"count": 923,
		"name": "Corsé"
	}, {
		"count": 612,
		"name": "Boisé"
	}, {
		"count": 1386,
		"name": "Fruité"
	}, {
		"count": 228,
		"name": "Doux"
	}, {
		"count": 94,
		"name": "Mousseux"
	}, {
		"count": 160,
		"name": "Léger"
	}, {
		"count": 54,
		"name": "Floral"
	}]
}, {
	"field": "tpaccordsnommenu",
	"label": "Suggested Pairings",
	"values": [{
		"count": 352,
		"name": "Agneau sur salsa d’herbes fraîches"
	}, {
		"count": 370,
		"name": "Boeuf barbecue à la marinade sèche"
	}, {
		"count": 215,
		"name": "Bavette au vin rouge, au poivre et au cumin"
	}, {
		"count": 244,
		"name": "Bobotie (hachis parmentier sud-africain)"
	}, {
		"count": 188,
		"name": "Brochettes de saucisses italiennes marinées"
	}, {
		"count": 149,
		"name": "Aubergines parmigiana"
	}, {
		"count": 64,
		"name": "Caponata sicilienne"
	}, {
		"count": 64,
		"name": "Cannellonis debout au boeuf et aux olives"
	}, {
		"count": 231,
		"name": "Boeuf braisé à la bière noire"
	}, {
		"count": 192,
		"name": "Burger de boeuf au cheddar et au bacon"
	}]
}, {
	"field": "tpobservationsgustativesacidite",
	"label": "Acidity",
	"values": [{
		"count": 2819,
		"name": "Rafraîchissante"
	}, {
		"count": 965,
		"name": "Vive"
	}, {
		"count": 186,
		"name": "Faible"
	}]
}, {
	"field": "tpobservationsgustativescorps",
	"label": "Body",
	"values": [{
		"count": 1981,
		"name": "Moyen"
	}, {
		"count": 1427,
		"name": "Corsé"
	}, {
		"count": 562,
		"name": "Léger"
	}]
}, {
	"field": "tpobservationsgustativessucre",
	"label": "Sweetness",
	"values": []
}, {
	"field": "tpobservationsgustativestannins",
	"label": "Tannins",
	"values": [{
		"count": 628,
		"name": "Souples"
	}, {
		"count": 1317,
		"name": "Charnus"
	}, {
		"count": 372,
		"name": "Fermes"
	}, {
		"count": 26,
		"name": "Rudes"
	}]
}, {
	"field": "tpobservationsgustativestexture",
	"label": "Texture",
	"values": [{
		"count": 2793,
		"name": "Ample"
	}, {
		"count": 494,
		"name": "Grasse"
	}, {
		"count": 178,
		"name": "Onctueuse"
	}, {
		"count": 505,
		"name": "Mince"
	}]
}];