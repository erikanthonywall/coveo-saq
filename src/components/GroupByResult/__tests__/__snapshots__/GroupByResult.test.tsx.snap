// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupByResult snapshot 1`] = `
<div
  className="noselect gbrContainer "
>
  <div
    className="gbrHeader"
  >
    <h3
      className="gbrFieldName"
      data-test-id="gbrFieldName"
    >
      Category
    </h3>
    <div
      className="collapseButton"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-chevron-up fa-w-14 collapseIcon"
        data-icon="chevron-up"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
  <div
    className="collapseContainer"
  >
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Vin rouge
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Vin blanc
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Vin rosé
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Champagne
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Rhum brun
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Whiskey américain
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Whisky écossais
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Vin mousseux
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Xérès
        </span>
      </label>
    </div>
    <div
      className="filterRow"
    >
      <label
        className="checkbox gbrCheckLabel"
      >
        <input
          checked={false}
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="gbrCheckLabelName"
          data-test-id="gbrCheckLabelName"
        >
          Dry gin
        </span>
      </label>
    </div>
  </div>
</div>
`;
